File: leetcode/0006-zigzag-conversion/solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        """
        RUNTIME: 36 ms (99.96%)
        MEMORY: 14MB (49.69%)
        """
        if numRows == 1: return s
        each = ['' for _ in range(numRows)]
        n = (numRows - 1) * 2
        for i in range(len(s)):
            index = i % n
            if index >= numRows:
                each[2 * (numRows - 1) - index] += s[i]
            else:
                each[index] += s[i]
        return ''.join(each)

File: leetcode/0006-zigzag-conversion/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 17 ms (58.81%)
     * MEMORY: 14.9 MB (22.45%)
    */
    string convert(string s, int numRows) {
        if (numRows == 1) { return s; }
        vector<string> each;
        for (size_t i=0; i<numRows; i++) {
            each.push_back("");
        }
        int n = (numRows - 1) * 2;
        for (size_t i=0; i<s.size(); i++) {
            int index = i % n;
            if (index >= numRows) {
                each[2 * (numRows - 1) - index] += s[i];
            }
            else {
                each[index] += s[i];
            }
        }
        string answer = "";
        for (string& str : each) {
            answer += str;
        }
        return answer;
    }
};
